// vibe-hnindex/dist/tools/watch-project.js
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { config } from '../config.js';
import { chunkFile } from '../services/chunker.js';
import { detectLanguage } from '../services/file-scanner.js';
import { embed } from '../services/embeddings.js';
import { getProject, insertChunks, deleteFileChunks, getExistingFileHash, updateProjectStats, getProjectFileCount, getProjectChunkCount, getAllProjectFiles, insertDependencies, insertExports, insertSymbols, deleteFileDependencies, deleteFileExports, deleteSymbolsForFile, } from '../services/sqlite.js';
import { parseImports, parseExports, resolveImportPath } from '../services/dependency-parser.js';
import { extractSymbols, toSymbolRecords } from '../services/symbol-extractor.js';
import { upsertPoints, deletePoints, healthCheck as qdrantHealthCheck, } from '../services/qdrant.js';
import { healthCheck as ollamaHealthCheck } from '../services/embeddings.js';
import { isIgnored, loadHnindexIgnore } from '../services/hnindex-ignore.js';
// Track active watchers per project
const activeWatchers = new Map();
// Debounce map to avoid re-indexing the same file multiple times
const pendingFiles = new Map();
const DEBOUNCE_MS = 1000;
// Cleanup all watchers on process exit
function cleanupAllWatchers() {
    for (const [name, entry] of activeWatchers) {
        entry.watcher.close();
        console.error(`[watch] Stopped watching: ${name}`);
    }
    activeWatchers.clear();
    for (const timeout of pendingFiles.values()) {
        clearTimeout(timeout);
    }
    pendingFiles.clear();
}
process.on('exit', cleanupAllWatchers);
process.on('SIGINT', () => { cleanupAllWatchers(); process.exit(0); });
process.on('SIGTERM', () => { cleanupAllWatchers(); process.exit(0); });
async function reindexFile(projectName, rootPath, absolutePath) {
    const relativePath = path.relative(rootPath, absolutePath).replace(/\\/g, '/');
    // Read file
    let content;
    try {
        const stat = fs.statSync(absolutePath);
        if (stat.size > config.maxFileSize || stat.size === 0)
            return null;
        const buffer = fs.readFileSync(absolutePath);
        // Binary check
        const checkLen = Math.min(buffer.length, 8192);
        for (let i = 0; i < checkLen; i++) {
            if (buffer[i] === 0)
                return null;
        }
        content = buffer.toString('utf-8');
    }
    catch {
        return null;
    }
    // Hash check
    const fileHash = crypto.createHash('sha256').update(content).digest('hex');
    const existingHash = getExistingFileHash(projectName, relativePath);
    if (existingHash === fileHash)
        return null; // unchanged
    // Delete old chunks
    if (existingHash !== null) {
        const oldIds = deleteFileChunks(projectName, relativePath);
        const qdrantOk = await qdrantHealthCheck();
        if (qdrantOk && oldIds.length > 0) {
            await deletePoints(projectName, oldIds);
        }
    }
    // Chunk + embed
    const language = detectLanguage(absolutePath);
    const chunks = chunkFile(content, relativePath);
    const now = new Date().toISOString();
    const records = chunks.map(chunk => ({
        id: crypto.randomUUID(),
        projectName,
        filePath: relativePath,
        absolutePath,
        chunkIndex: chunk.chunkIndex,
        startLine: chunk.startLine,
        endLine: chunk.endLine,
        content: chunk.content,
        language,
        fileHash,
        indexedAt: now,
    }));
    const ollamaOk = await ollamaHealthCheck();
    if (ollamaOk && records.length > 0) {
        try {
            const vectors = await embed(records.map(r => r.content));
            insertChunks(records);
            const qdrantOk = await qdrantHealthCheck();
            if (qdrantOk) {
                const points = records.map((r, i) => ({
                    id: r.id,
                    vector: vectors[i],
                    payload: {
                        project_name: r.projectName,
                        file_path: r.filePath,
                        chunk_index: r.chunkIndex,
                        start_line: r.startLine,
                        end_line: r.endLine,
                        language: r.language,
                    },
                }));
                await upsertPoints(projectName, points);
            }
        }
        catch (error) {
            console.error(`[watch] Embed error for ${relativePath}:`, error);
            insertChunks(records);
        }
    }
    else {
        insertChunks(records);
    }
    try {
        deleteFileDependencies(projectName, relativePath);
        deleteFileExports(projectName, relativePath);
        deleteSymbolsForFile(projectName, relativePath);
        const projectFiles = getAllProjectFiles(projectName);
        const imports = parseImports(content, language);
        const depRecords = imports.map(imp => ({
            id: crypto.randomUUID(),
            projectName,
            sourceFile: relativePath,
            targetFile: resolveImportPath(imp.specifier, relativePath, projectFiles) ?? imp.specifier,
            importSpecifiers: imp.specifiers,
            importType: imp.importType,
            language,
        }));
        if (depRecords.length > 0)
            insertDependencies(depRecords);
        const parsedExports = parseExports(content, language);
        const exportRecords = parsedExports.map(exp => ({
            id: crypto.randomUUID(),
            projectName,
            filePath: relativePath,
            exportName: exp.name,
            exportType: exp.exportType,
            lineNumber: exp.lineNumber,
            language,
        }));
        if (exportRecords.length > 0)
            insertExports(exportRecords);
        const parsedSymbols = extractSymbols(content, language);
        const symbolRecords = toSymbolRecords(projectName, relativePath, language, parsedSymbols);
        if (symbolRecords.length > 0)
            insertSymbols(symbolRecords);
    }
    catch (parseError) {
        console.error(`[watch] Dependency/symbol parse failed for ${relativePath}:`, parseError);
    }
    // Update stats
    const fileCount = getProjectFileCount(projectName);
    const chunkCount = getProjectChunkCount(projectName);
    updateProjectStats(projectName, fileCount, chunkCount);
    return relativePath;
}
// Supported extensions for watch filtering
const WATCH_EXTENSIONS = new Set([
    '.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.py', '.pyi', '.java', '.go',
    '.rs', '.c', '.cpp', '.cc', '.cxx', '.h', '.hpp', '.hxx', '.cs', '.rb', '.php',
    '.swift', '.kt', '.kts', '.scala', '.lua', '.sh', '.bash', '.zsh', '.sql',
    '.vue', '.svelte', '.html', '.htm', '.css', '.scss', '.less', '.sass',
    '.yaml', '.yml', '.toml', '.json', '.xml', '.md', '.mdx',
    '.proto', '.graphql', '.gql', '.tf', '.zig', '.ex', '.exs', '.erl',
    '.clj', '.cljs', '.hs', '.ml', '.mli', '.fs', '.fsx', '.dart', '.sol',
]);
const SKIP_DIRS = new Set([
    'node_modules', '.git', 'dist', 'build', 'out', '.next', '__pycache__',
    '.venv', 'venv', 'target', 'vendor', 'coverage', '.cache',
]);
function pathHasEggInfoSegment(filename) {
    return filename.replace(/\\/g, '/').split('/').some(p => p.endsWith('.egg-info'));
}
export async function startWatchingProject(projectName) {
    const project = getProject(projectName);
    if (!project) {
        return {
            ok: false,
            message: `Project "${projectName}" not found. Run index_codebase first.`,
        };
    }
    if (activeWatchers.has(projectName)) {
        return {
            ok: true,
            message: `Already watching project "${projectName}" at ${project.rootPath}`,
        };
    }
    const rootPath = project.rootPath;
    const resolvedRoot = path.resolve(rootPath);
    const ignorePatterns = loadHnindexIgnore(resolvedRoot);
    const watcher = fs.watch(rootPath, { recursive: true }, (eventType, filename) => {
        if (!filename)
            return;
        const fullPath = path.join(rootPath, filename);
        const ext = path.extname(filename).toLowerCase();
        if (!WATCH_EXTENSIONS.has(ext))
            return;
        const parts = filename.replace(/\\/g, '/').split('/');
        if (parts.some(p => SKIP_DIRS.has(p)))
            return;
        if (pathHasEggInfoSegment(filename))
            return;
        const existing = pendingFiles.get(fullPath);
        if (existing)
            clearTimeout(existing);
        pendingFiles.set(fullPath, setTimeout(async () => {
            pendingFiles.delete(fullPath);
            const relativePath = path.relative(resolvedRoot, fullPath).replace(/\\/g, '/');
            if (!fs.existsSync(fullPath)) {
                const oldIds = deleteFileChunks(projectName, relativePath);
                deleteFileDependencies(projectName, relativePath);
                deleteFileExports(projectName, relativePath);
                deleteSymbolsForFile(projectName, relativePath);
                if (oldIds.length > 0) {
                    const qdrantOk = await qdrantHealthCheck();
                    if (qdrantOk)
                        await deletePoints(projectName, oldIds);
                }
                console.error(`[watch] Removed: ${relativePath}`);
                return;
            }
            if (isIgnored(relativePath, ignorePatterns))
                return;
            const result = await reindexFile(projectName, resolvedRoot, fullPath);
            if (result) {
                console.error(`[watch] Re-indexed: ${result}`);
            }
        }, DEBOUNCE_MS));
    });
    activeWatchers.set(projectName, { watcher, count: 0 });
    return {
        ok: true,
        message: [
            `Watching project "${projectName}" for file changes.`,
            `Path: ${rootPath}`,
            `Files will be automatically re-indexed when saved.`,
            `Use unwatch_project to stop watching.`,
        ].join('\n'),
    };
}
export async function watchProjectTool(args) {
    const result = await startWatchingProject(args.project_name);
    return {
        content: [{ type: 'text', text: result.message }],
    };
}
export async function unwatchProjectTool(args) {
    const entry = activeWatchers.get(args.project_name);
    if (!entry) {
        return {
            content: [{
                    type: 'text',
                    text: `Project "${args.project_name}" is not being watched.`,
                }],
        };
    }
    entry.watcher.close();
    activeWatchers.delete(args.project_name);
    return {
        content: [{
                type: 'text',
                text: `Stopped watching project "${args.project_name}".`,
            }],
    };
}
